### PR TITLE
Fix disabling a module for a specific shop applied to all shops

### LINKED ISSUE #30262: Disabling a module in one shop will disable it in all shops
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

I find this regression while testing these PRs: #29516 & https://github.com/PrestaShop/classic-theme/pull/52 & https://github.com/PrestaShop/ps_contactinfo/pull/49 & https://github.com/PrestaShop/hummingbird/pull/366

Enable the multistore then go to Module Manager page, try to click on configure button of any module. 

1. You will find a trivial regression :smile: (see the strong tag displayed while it should not be displayed)
2. Now, uncheck the **Activate module for this shop context** option to disable the module for this store. Check all stores, you will see that the module is disabled for all shops not just the selected shop. (see the attached screenshot below)

 
image

### Expected behavior

1. The strong tag should be displayed
2. The module should be disabled only in the selcted store

### Steps to reproduce

1. BO>CONFIGURE>Shop Parameters>General
2. Enable Multistore then Save
3. BO>CONFIGURE>Advanced Parameters>Multistore
4. Create a second store
5. BO>IMPROVE>Modules>Module Manager
6. Click on Configure button of ps_contactinfo module (for exemple)
7. In Configuration bloc
8. See error > the strong tag displayed
9. Uncheck the **Activate module for this shop context** option to disable the module for the selected shop
10. See error > Check all stores, you will see that the module is disabled for all shops not just the selected shop

### PrestaShop version(s) where the bug happened

8.0.0

### PHP version(s) where the bug happened

7.2

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | when disabling a module, the query must specify id shops if we don't want to disable the module for all shops
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      |  See steps in issue #30262 Also make sure that the change doesn't have side effects when the shop didn't enable multishop (enable/disable module feature should still work properly)
| Fixed ticket?     | Fixes #30262 / #32211 (to be verified)
| Related PRs       | none
| Sponsor company   | PrestaShop SA

### CODE REVIEW

File: classes/module/Module.php [hunk 1]
```diff
@@ -983,8 +983,9 @@ public function disable($force_all = false)
             $result &= $this->uninstallOverrides();
         }
 
-        // Disable module for all shops
-        $result &= Db::getInstance()->delete('module_shop', '`id_module` = ' . (int) $this->id);
+        // Disable module for all shops or contextual shops
+        $whereIdShop = $force_all ? '' : ' AND `id_shop` IN(' . implode(', ', Shop::getContextListShopID()) . ')';
```
Comment (PululuK) [inline classes/module/Module.php:987]: can you please check if multi-shop feature is enabled ?

Comment (matthieu-rolland) [reply classes/module/Module.php:987]: This is a legacy function and we never made this check before (I'm reverting the query here), I thought about it and realized that it's actually not necessary.

Did I miss a use case where it makes sense ?

**Review by Hlavtox — DISMISSED** [review]
Looks good

**Review by atomiix — DISMISSED** [review]

**Review by hibatallahAouadni — COMMENTED** [review]
Hello @matthieu-rolland 

Thanks for your PR :pray: 
Unfortunately, there's a regression on your PR :disappointed: 
When we disable the module by this option, the module is disabled but the Disable button/option in select still displayed, while it shouldn't!
It should display the Enable button/option in the select :wink: 
See the attached screen record below:

[pr_31379.webm](https://user-images.githubusercontent.com/23308427/220335292-b9ec8bff-fd5e-4fc4-be23-bf226fddeac1.webm)

Please check and feedback.

Thanks!

**Comment by kpodemski** [comment]
@matthieu-rolland @PrestaShop/product-team, what about this PR? It fixes a major issue

File: classes/module/Module.php [hunk 2]
```diff
@@ -983,8 +983,9 @@ public function disable($force_all = false)
...
+        $result &= Db::getInstance()->delete('module_shop', '`id_module` = ' . (int) $this->id . $whereIdShop);
```
Comment (jolelievre) [inline classes/module/Module.php:988]: No changes are needed in `enable` method right? It was just not correctly done for disable?

**Review by jolelievre — COMMENTED** [review]
Thanks @matthieu-rolland 

I searched for other usages of `isActive` just to be sure:
- https://github.com/PrestaShop/PrestaShop/blob/840fc123126857f93925598d47b5d86c96edf35d/src/PrestaShopBundle/EventListener/ModuleActivatedListener.php#L122 It should probably be handled here as well, especially since this is the listener that automatically redirects when module is activated or not
- https://github.com/PrestaShop/PrestaShop/blob/c52e131b5e446d3d414701835fe8c288dfe702f8/src/Core/Module/ModuleRepository.php#L122 I think this is the part that checks id configure action is allowed, in this case it makes sense to check for the global active state not the multishop one so this one is fine

Just one thing that worries me, I couldn't find the usage in AdminModuleDataProvider via PHPStorm probably because it's based on a ModuleCollection and the loop variable is not annotated, which is not that much of a problem in itself but we should be careful that there are no other "hidden usage" at least by checking other usage of the ModuleCollection I guess.

File: classes/module/Module.php (hunk 2 above)
Comment (matthieu-rolland) [reply classes/module/Module.php:988]: yep

**Review by mflasquin — DISMISSED** [review]

**Review by 0x346e3730 — DISMISSED** [review]

File: src/Adapter/Module/ModuleDataProvider.php [hunk 3]
```diff
@@ -370,4 +368,18 @@ private function getDeviceStatus($name)
...
+
+    private function isModuleActive(int $id, array $id_shops): ?array
+    {
+        $result = Db::getInstance()->getRow('SELECT m.`id_module` as `active`, ms.`id_module` as `shop_active`, ms.`enable_device` as `enable_device`
+            FROM `' . _DB_PREFIX_ . 'module` m
+            LEFT JOIN `' . _DB_PREFIX_ . 'module_shop` ms ON m.`id_module` = ms.`id_module`
+            WHERE m.`id_module` = ' . $id . '
+            AND ms.`id_shop` IN (' . implode(',', $id_shops) . ')');
+        if ($result) {
+            return $result;
```
Comment (jolelievre) [inline src/Adapter/Module/ModuleDataProvider.php]: It's a bit weird you return an associative array of value but never check their content, the returned type is an array and is cast as a boolean Shouldn't the method return a boolean instead?

Comment (jolelievre) [reply src/Adapter/Module/ModuleDataProvider.php]: Something like:
```suggestion
            return !empty($result['enable_device']) && !empty($result['active']);
```

**Review by jolelievre — COMMENTED** [review]
Thanks @FabienPapet 

So you completely dropped the new getter and instead, you provide dynamic data in the parameter bag based on the current shop context right? Seems like a good idea However, the `isModuleActive` doesn't see to be return exactly what we need 🤔 (see comments)

File: src/Adapter/Module/ModuleDataProvider.php [hunk 4]
```diff
@@ -160,9 +160,7 @@ public function getInstalled(): array
             if (array_key_exists('active_on_mobile', $module)) {
                 $module['active_on_mobile'] = (bool) ($module['active_on_mobile'] & AddonListFilterDeviceStatus::DEVICE_MOBILE);
             }
-            if (array_key_exists('active', $module)) {
-                $module['active'] = (bool) $module['active'];
-            }
+            $module['active'] = (bool) $this->isModuleActive($module['id'], $id_shops);
```
Comment (jolelievre) [inline src/Adapter/Module/ModuleDataProvider.php]:
```suggestion
            $module['active'] = !empty($result['enable_device']) && !empty($result['active']);
```

File: src/Adapter/Module/ModuleDataProvider.php [hunk 5]
```diff
@@ -146,7 +146,7 @@ public function getInstalled(): array
 
         $id_shops = (new Context())->getContextListShopID();
         if (count($id_shops) > 0) {
-            $select .= ', ms.`id_module` as active, ms.`enable_device` as active_on_mobile';
+            $select .= ', ms.`enable_device` as active_on_mobile';
```
Comment (jolelievre) [inline src/Adapter/Module/ModuleDataProvider.php:149]:
```suggestion
            $select .= ', ms.`active` as active, ms.`enable_device` as active_on_mobile';
```

**Review by jolelievre — APPROVED** [review]
Thanks @matthieu-rolland and @FabienPapet

**Review by boherm — APPROVED** [review]
LGTM @matthieu-rolland!

**Review by hibatallahAouadni — COMMENTED** [review]
Hello @matthieu-rolland 

Thanks for your hard working :muscle: 
Well, it seems that the mentioned issue with the `GDPR` module version `2.0.1` is fixed :tada: 
Even thought, I needed to download psgdpr_2.0.1 to upgrade my psgdpr :sweat_smile: (it seems when we release a module with label pre-release, PS doesn't take it into account).
Then, I checked this issue #30262 and it seems fixed :tada: here's the test cases that I use:

- Disable the module with the **checkbox** in Configure page with **All stores**: :heavy_check_mark: the module is not displayed in FO
- Disable the module with the button **Disable** in the Module Manager page with **All stores**: :heavy_check_mark: the module is well disabled and not displayed in FO
- Disable the module with the **checkbox** in Configure page in **Shop1** => :heavy_check_mark: the module is not displayed in FO Shop1 and it's displayed in FO Shop2
- Disable the module with the button **Disable** in the Module Manager page in **Shop1** => :heavy_check_mark: the module is not displayed in FO Shop1 and it's displayed in FO Shop2
- Disable the module with the **checkbox** in Configure page in **Shop2** => :heavy_check_mark: the module is not displayed in FO Shop2 and it's displayed in FO Shop1
- Disable the module with the button **Disable** in the Module Manager page in **Shop2** => :heavy_check_mark: the module is not displayed in FO Shop2 and it's displayed in FO Shop1

But, the trivial regression of displaying the tag 'strong' still persist, see the attached screenshot below:

image

Otherwise, the auto tests are :green_circle: https://github.com/aniszr/ga.tests.ui.pr/actions/runs/5199173629

Waiting for your feedback.

Thanks!

**Comment by MatShir** [comment]
@hibatallahAouadni for the version, It was made in purpose bc the new version was breaking upgrades. Now everything seems to be fixed.

**Comment by MatShir** [comment]
Thanks everyone, nice work 🥳 . Let's fix the "< strong >" issue on another since we have QA and Dev approved, wdyt @hibatallahAouadni & @matthieu-rolland ?

**Comment by jolelievre** [comment]
Thanks @matthieu-rolland and @FabienPapet

**Comment by matks** [comment]
Thank you everyone 😉

**Comment by RosaBenouamer** [comment]
Congratulations to the team :-)

**Pull request merged**